pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -0,0 +1,769 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package memorymanager
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	corehelper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/memorymanager/state"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+)
+
+const policyTypeStatic policyType = "Static"
+
+type systemReservedMemory map[int]map[v1.ResourceName]uint64
+
+// SingleNUMAPolicy is implementation of the policy interface for the single NUMA policy
+type staticPolicy struct {
+	// machineInfo contains machine memory related information
+	machineInfo *cadvisorapi.MachineInfo
+	// reserved contains memory that reserved for kube
+	systemReserved systemReservedMemory
+	// topology manager reference to get container Topology affinity
+	affinity topologymanager.Store
+}
+
+var _ Policy = &staticPolicy{}
+
+// NewPolicyStatic returns new single NUMA policy instance
+func NewPolicyStatic(machineInfo *cadvisorapi.MachineInfo, reserved systemReservedMemory, affinity topologymanager.Store) (Policy, error) {
+	var totalSystemReserved uint64
+	for _, node := range reserved {
+		if _, ok := node[v1.ResourceMemory]; !ok {
+			continue
+		}
+		totalSystemReserved += node[v1.ResourceMemory]
+	}
+
+	// check if we have some reserved memory for the system
+	if totalSystemReserved <= 0 {
+		return nil, fmt.Errorf("[memorymanager] you should specify the system reserved memory")
+	}
+
+	return &staticPolicy{
+		machineInfo:    machineInfo,
+		systemReserved: reserved,
+		affinity:       affinity,
+	}, nil
+}
+
+func (p *staticPolicy) Name() string {
+	return string(policyTypeStatic)
+}
+
+func (p *staticPolicy) Start(s state.State) error {
+	if err := p.validateState(s); err != nil {
+		klog.Errorf("[memorymanager] Invalid state: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	return nil
+}
+
+// Allocate call is idempotent
+func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) error {
+	// allocate the memory only for guaranteed pods
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
```

Comment:
so, it was kind of decided a year ago that Topology Alignment should not be tied to the QoS class of a Pod, see: https://github.com/kubernetes/kubernetes/pull/83492

are we going back on this and placing an extra restriction on memory management, and if yes, why?

Reply:
It pretty the same for the CPU manager - https://github.com/kubernetes/kubernetes/blob/b2ecd1b3a3192fbbe2b9e348e095326f51dc43dd/pkg/kubelet/cm/cpumanager/policy_static.go#L220, and the reason that only for the guaranteed class you can reserve the exact amount of CPUs or memory. In general, it should not affect the topology manager, for all other QoS classes the hint will be nil, so from the topology manager every hint will be good for the hint provider.

Reply:
the reason while it makes sense in the CPU Manager to do that, is because exclusive CPU allocation, aka. the only resource managed by that component possibly needing NUMA alignment is already tied to the Guaranteed QoS class. if the Pod is not in Guaranteed QoS for sure it did not ask for exclusive CPUs, so for sure there is nothing to align, easy early return case

the allocation of hugepages on the other hand -the resources managed by this manager possibly needing NUMA alignment- is not tied to Guaranteed QoS class. the documentation states the following: "Note that when requesting hugepage resources, either memory **or** CPU resources must be requested as well."
Even the example Pod is not in the Guaranteed QoS class (https://kubernetes.io/docs/tasks/manage-hugepages/scheduling-hugepages/)

so, applying the same restriction to a resource working differently goes against the intention of the PR I linked. the reason why that restriction was removed from the topology manager is because not all workloads requiring NUMA allocation asks for exclusive CPUs, or will be in the guaranteed QoS class. don't forget, topology manager manages three resource types. isn't a possible an app just wants hugepages, and devices?

so, what is the functional reason behind having the restriction in place? if it was simply an assumption that it is needed based on the CPU manager code but there is no explicit reason to have it, IMO it should be removed from the code
if the functional reason is to be able to early return in the majority of the use-case, just simply apply the logic -but not the code- of the CPU Manager early return case. "do I have a possibly NUMA alignable resource I might need to take care of, or can I stop processing?"
unlike with CPU Manager this doesn't require implicit guessing though. you could just literally check if hugepages were requested, and return if not

Reply:
> so, what is the functional reason behind having the restriction in place? if it was simply an assumption that it is needed based on the CPU manager code but there is no explicit reason to have it, IMO it should be removed from the code
> if the functional reason is to be able to early return in the majority of the use-case, just simply apply the logic -but not the code- of the CPU Manager early return case. "does this Pod even have a possibly NUMA alignable resource I might need to take care of, or can I stop processing?"
> unlike CPU Management verifying this doesn't require implicit guessing though. you could just literally check if hugepages were requested, and return if not

Like I wrote above, the main reason it only for guaranteed class we can calculate correctly the amount of reserved and free memory. Another reason, that only for the guaranteed containers we can guarantee the maximal density of containers, please see [link](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager#simulation---how-the-memory-manager-works-by-examples).

Reply:
memory request = memory limits and Guaranteed QoS are not the same though, and that's my main issue
Guaranteed QoS also means **CPU** requests and limits 1: exist 2: are equal
To clarify, I'm not worried about admittance. I'm worried about alignment. When you return nil to the Topology Manager in all non-guaranteed cases you also push all the Pods into preferred. And in some cases I really don't think that's correct

Let's run it through an example. Take the case of a container having 5GB hugepages in both request and limit, and also asking for one Device in both requests and limits. This container won't be guaranteed because of the missing CPU request/limits, right?
So with the current code memory manager returns "all NUMA are ok", CPU manager returns "all NUMA are ok". Device Manager let's say returns "NUMA [0,1] are ok" because there are available Devices from these nodes. 

What happens in this case if memory in reality would be available from both nodes? The Pod is admitted even into a a single-NUMA policy node, and the memory will be assigned disjointed from the Device assignment, right? basically alignment will be up to luck, again, on a Node having single-numa policy where the requesting container rightfully expected to always get the memory and the Device from the same place

if I'm wrong please feel free to correct me :) if not, I understand the need to have a check, I only argue that it should not be checking Guaranteed QoS, but maybe checking request = limit purely for memory, without implicitly involving the CPU requests and limits via Guaranteed QoS

Reply:
A lot of thanks for the example, it totally makes sense. 

An additional reason why we want to allocate memory only for guaranteed pods because we want to prevent additional OOM errors, for example, you run a lot of burstable and best-effort containers on the NUMA node 0, and you start an additional burstable container that has hugepages requests and limits equal, so the memory manager will check the memory reservation and will see that the NUMA node 0 has enough memory(we did not add previous container reservations because requests.hugepages != limits.hugepages) and will pin the container to the NUMA node 0, but it does not really have enough memory because it already consumed by previous containers and our new container can be killed because of the OOM(it has the same `oom_score_adj` as all other containers) and it can happen again and again. 

The guaranteed container has a smaller `oom_score_adj` value(https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#node-oom-behavior) so it will be killed only after Burstable and Best-Efforts containers, but these containers can just be restarted on the different NUMA node(no pinning for them). 

I hope it will clarify why do we apply static policy logic to the guaranteed pods.

Reply:
yeah, this totally makes sense! thanks for the extra elaboration.
this way my -unfortunetaly not-so hypotethical- scenario would be still unaddressed, so I wonder if there is a way to reconcile the two and find a common working solution?

Maybe:
- setting/overwriting the OOM score of "NUMA aligned containers" regardless of their QoS class would solve the issue right? indeed,  it could result in the removal of other non-pinned containers from the same node but as you mentioned that is generally not seen as problematic, cause it would happen automatically
- bookkeeping the hugepages **limits** of all containers, and make decisions based on that? it is kind of dimensioning for the worst-case, true, but something for something

I'm totally fine discussing this outside the scope of this PR, but can we at least add this issue into the scope of the beta plan? just to sensure we spend some cycles thinking about it, maybe coming up with a configuration option if a generic allgorithm cannot be found etc.

Reply:
@Levovar Totally makes sense, let's continue the discussion under the [document](https://docs.google.com/document/d/1FHvJgleXkyIWiWfntbneF-cYdms7sJ2SiDj3yfRfqXA/edit#heading=h.fhko4rmz59bh)